### ISSUE TITLE
Configurable capacity UniqueToken

### ISSUE THREAD

**Issue by hcedwar** [kokkos/kokkos#979]
Request by @crtrott 

The UniqueToken capability has hard-wired capacity of the maximum concurrency of the underlying execution space.  This functionality assumes that all concurrent threads will request / release tokens, such as when using a RangePolicy.

An enhancement:  Let the default capacity remain the maximum concurrency. Also allow the capacity to be configurable.  Use case is tokens requested once per team using a TeamPolicy.

```c++
template< class ExecSpace >
class UniqueToken {
public:
  static int default_capacity();
  UniqueToken( ExecSpace const & = ExecSpace() , int capacity = default_capacity() );
  // ... etc.
};
```

**Comment by hcedwar** [comment]
Issue: The host implementations do not use a bitset that can be given a size.
Will need a different type to use this.
Perhaps the `ConcurrentBitSet` could be used directly.
Follow up with @crtrott and @mhoemmen

**Comment by mhoemmen** [comment]
@hcedwar Sure, this would be handy for thread teams or for threads within a task.

**Comment by dsunder** [comment]
Add Dynamic template

**Comment by mhoemmen** [comment]
@dsunder just curious -- why do you want to template it?  is it faster if not a run-time parameter?

**Comment by ibaned** [comment]
@mhoemmen yes its faster, currently all the CPU-target implementations query some kind of thread ID function (like OpenMP), whereas the "Dynamic" mode requires you to use a data structure like the Kokkos::Bitset.

**Comment by mhoemmen** [comment]
As of last week, work with @vbrunini et al. has resulted in a prototype that fixes both issues mentioned in #1806.  Here are the tasks that remain:

- [ ] Port to back-ends other than Cuda, OpenMP, and Serial
- [ ] Add tests for AcquireUniqueToken (RAII helper)
- [ ] Possibly add back in the deprecated interfaces (or not?)

**Comment by jgfouca** [comment]
+1 for this functionality.
